community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/cardinality/assumeIndependence/PatternSelectivityCalculator.scala

```diff
@@ -42,24 +42,26 @@ case class PatternSelectivityCalculator(stats: GraphStatistics, combiner: Select
 
   def apply(pattern: PatternRelationship, labels: Map[IdName, Set[LabelName]])
            (implicit semanticTable: SemanticTable, selections: Selections): Selectivity = {
-    val allNodes = stats.nodesWithLabelCardinality(None)
+    val nbrOfNodesInGraph = stats.nodesWithLabelCardinality(None)
     val lhs = pattern.nodes._1
     val rhs = pattern.nodes._2
     val labelsOnLhs: Seq[TokenSpec[LabelId]] = mapToLabelTokenSpecs(selections.labelsOnNode(lhs) ++ labels.getOrElse(lhs, Set.empty))
     val labelsOnRhs: Seq[TokenSpec[LabelId]] = mapToLabelTokenSpecs(selections.labelsOnNode(rhs) ++ labels.getOrElse(rhs, Set.empty))
 
-    val lhsCardinality = allNodes * calculateLabelSelectivity(labelsOnLhs)
-    val rhsCardinality = allNodes * calculateLabelSelectivity(labelsOnRhs)
+    val lhsCardinality = nbrOfNodesInGraph * calculateLabelSelectivity(labelsOnLhs, nbrOfNodesInGraph)
+    val rhsCardinality = nbrOfNodesInGraph * calculateLabelSelectivity(labelsOnRhs, nbrOfNodesInGraph)
 
     // If either side of our pattern is empty, it's all empty
-    if (lhsCardinality == Cardinality.EMPTY || lhsCardinality == Cardinality.EMPTY)
-      Selectivity(1)
+    if (lhsCardinality == Cardinality.EMPTY || rhsCardinality == Cardinality.EMPTY)
+      Selectivity.ZERO
```

Comment:
Yes, selectivity is not cardinality. But the _value_ for the selectivity in our code is directly proportional to cardinality. This means that selectivity zero => cardinality zero. If a pattern has no possible endpoint (for one of the endpoints), then surely the cardinality of the pattern should be zero, right?

`MATCH (a:A)-->(b:B)` and there are no `:B`-nodes, then surely there can be no matches.